src/main/java/org/springframework/data/redis/connection/ClusterCommandExecutor.java

```diff
@@ -0,0 +1,352 @@
+
+		if (redirectCount > maxRedirects) {
+			throw new TooManyClusterRedirectionsException(
+					String
+							.format(
+									"Cannot follow Cluster Redirects over more than %s legs. Please consider increasing the number of redirects to follow. Current value is: %s.",
+									redirectCount, maxRedirects));
+		}
+
+		S client = this.resourceProvider.getResourceForSpecificNode(node);
+		Assert.notNull(client, "Could not acquire resource for node. Is your cluster info up to date?");
+
+		try {
+			return cmd.doInCluster(client);
+		} catch (RuntimeException ex) {
+
+			RuntimeException translatedException = convertToDataAccessExeption(ex);
+			if (translatedException instanceof ClusterRedirectException) {
+				ClusterRedirectException cre = (ClusterRedirectException) translatedException;
+				return executeCommandOnSingleNode(cmd,
+						topologyProvider.getTopology().lookup(cre.getTargetHost(), cre.getTargetPort()), redirectCount + 1);
+			} else {
+				throw translatedException != null ? translatedException : ex;
+			}
+		} finally {
+			this.resourceProvider.returnResourceForSpecificNode(node, client);
+		}
+	}
+
+	/**
+	 * Run {@link ClusterCommandCallback} on all reachable master nodes.
+	 * 
+	 * @param cmd
+	 * @return
+	 * @throws ClusterCommandExecutionFailureException
+	 */
+	public <S, T> Map<RedisClusterNode, T> executeCommandOnAllNodes(final ClusterCommandCallback<S, T> cmd) {
+		return executeCommandAsyncOnNodes(cmd, getClusterTopology().getActiveMasterNodes());
+	}
+
+	/**
+	 * @param callback
+	 * @param nodes
+	 * @return
+	 * @throws ClusterCommandExecutionFailureException
+	 */
+	public <S, T> java.util.Map<RedisClusterNode, T> executeCommandAsyncOnNodes(
+			final ClusterCommandCallback<S, T> callback, Iterable<RedisClusterNode> nodes) {
+
+		Assert.notNull(callback, "Callback must not be null!");
+		Assert.notNull(nodes, "Nodes must not be null!");
+
+		Map<RedisClusterNode, Future<T>> futures = new LinkedHashMap<RedisClusterNode, Future<T>>();
+		for (final RedisClusterNode node : nodes) {
+
+			futures.put(node, executor.submit(new Callable<T>() {
+
+				@Override
+				public T call() throws Exception {
+					return executeCommandOnSingleNode(callback, node);
+				}
+			}));
+		}
+
+		return collectResults(futures);
+	}
+
+	private <T> Map<RedisClusterNode, T> collectResults(Map<RedisClusterNode, Future<T>> futures) {
+
+		boolean done = false;
+
+		Map<RedisClusterNode, T> result = new HashMap<RedisClusterNode, T>();
+		Map<RedisClusterNode, Throwable> exceptions = new HashMap<RedisClusterNode, Throwable>();
+		while (!done) {
+
+			done = true;
+			for (Map.Entry<RedisClusterNode, Future<T>> entry : futures.entrySet()) {
+
+				if (!entry.getValue().isDone() && !entry.getValue().isCancelled()) {
+					done = false;
+				} else {
+					if (!result.containsKey(entry.getKey()) && !exceptions.containsKey(entry.getKey())) {
+						try {
+							result.put(entry.getKey(), entry.getValue().get());
```

Comment:
Awaiting a `Future` without a timeout can block forever. You get here only if the `Future` is done but the outer loop is also not protected by a timeout.